File: src/quantum/advanced/fractal_css_expanded.py
# EXPERIMENT 4B -- Expand CSS stabilizer families (12 X, 12 Z), compute ranks and search for a low-weight logical
# Strategy: Use Sierpinski mask; build 12 translations each; compute H_X, H_Z ranks; search for X-logical
# among even-parity translates (coarse grid). Render candidate support if found.
# Dark mode, 3160x2820.

import matplotlib.pyplot as plt
import numpy as np

# Rendering
W, H = 3160, 2820; dpi=100; fig_w, fig_h=W/dpi, H/dpi
plt.rcParams.update({
    "figure.facecolor":"#0d0f14",
    "axes.facecolor":"#0d0f14",
    "axes.edgecolor":"#1f2937",
    "axes.labelcolor":"white",
    "xtick.color":"white",
    "ytick.color":"white",
    "text.color":"white",
    "font.size":16
})

L=33
S = np.zeros((L,L), dtype=np.uint8)
S[0, L//2] = 1
for r in range(1, L):
    left = np.roll(S[r-1], 1, axis=0)
    right = np.roll(S[r-1], -1, axis=0)
    S[r] = left ^ right

coords = np.add.outer(np.arange(L), np.arange(L))
X_parity = ((coords % 2) == 0).astype(np.uint8)
Z_parity = 1 - X_parity

def translate(mask, tx, ty):
    return np.roll(np.roll(mask, ty, axis=0), tx, axis=1)

# 12 shifts each (even steps to preserve parity split)
X_shifts = [(2*a, 2*b) for a,b in [(0,0),(1,0),(0,1),(1,1),(2,1),(1,2),(3,0),(0,3),(3,2),(2,3),(4,1),(1,4)]]
Z_shifts = [(2*a, 2*b) for a,b in [(1,0),(0,1),(2,0),(0,2),(3,1),(1,3),(2,2),(4,0),(0,4),(3,3),(4,2),(2,4)]]

X_stabs = [(translate(S, tx, ty) & X_parity) for (tx,ty) in X_shifts]
Z_stabs = [(translate(S, tx, ty) & Z_parity) for (tx,ty) in Z_shifts]

n = L*L
def flat(m): return m.reshape(-1) % 2
HX = np.array([flat(m) for m in X_stabs], dtype=np.uint8)
HZ = np.array([flat(m) for m in Z_stabs], dtype=np.uint8)

# GF(2) rank
def gf2_rank(A):
    A = A.copy().astype(np.uint8)
    m,n = A.shape
    r = 0
    for c in range(n):
        piv = None
        for i in range(r, m):
            if A[i,c]:
                piv = i; break
        if piv is None:
            continue
        if piv != r:
            A[[r,piv]] = A[[piv,r]]
        for i in range(m):
            if i != r and A[i,c]:
                A[i] ^= A[r]
        r += 1
        if r == m: break
    return r

print("Computing GF(2) Ranks for Fractal Code...")
rank_HX = gf2_rank(HX)
rank_HZ = gf2_rank(HZ)
k = n - rank_HX - rank_HZ

# Candidate X-logical search among even-parity translates (coarse step=2)
best = None
for tx in range(0, L, 2):
    for ty in range(0, L, 2):
        xmask = translate(S, tx, ty) & X_parity
        xrow = flat(xmask)
        # must commute with all Z stabilizers
        if np.any((xrow @ HZ.T) % 2):
            continue
        # must be independent of H_X
        HX_ext = np.vstack([HX, xrow])
        if gf2_rank(HX_ext) == rank_HX:
            continue
        w = int(xrow.sum())
        if (best is None) or (w < best[0]):
            best = (w, tx, ty, xmask)

if best is None:
    print("No candidate X-logical found in this restricted family.")
else:
    w_best, tx_best, ty_best, Xlog = best
    print(f"Found X-Logical: w={w_best}, shift=({tx_best},{ty_best})")
    # Render
    img = np.zeros((L, L, 3), dtype=float)
    blue = np.array([0.38, 0.65, 0.96])
    img += Xlog[..., None] * blue
    fig = plt.figure(figsize=(fig_w, fig_h), dpi=dpi)
    ax = plt.gca()
    ax.imshow(img, origin="upper", interpolation="nearest")
    ax.set_title("Fractal CSS (Expanded): Candidate X-logical Support (Blue)\n"
                 f"L={L}, n={n}, rank(HX)={rank_HX}, rank(HZ)={rank_HZ}, k={k}, weight={w_best}",
                 fontsize=28, loc="left")
    ax.set_xlabel("Lattice X index"); ax.set_ylabel("Lattice Y index")
    ax.set_xticks([]); ax.set_yticks([])
    plt.tight_layout()
    outp = "data/artifacts/images/css_fractal_candidate_logical_expanded_3160x2820.png"
    plt.savefig(outp, dpi=dpi, facecolor="#0d0f14")
    plt.close()
    print("Saved:", outp)

File: crates/gororoba_cli_quantum/src/bin/fractal-css-expanded.rs
use anyhow::Result;
use clap::Parser;
use image::{ImageBuffer, Rgb};
use std::path::PathBuf;

#[derive(Parser, Debug)]
#[command(author, version, about = "Fractal CSS Expanded Stabilizer Search")]
struct Args {
    #[arg(short, long, default_value = "data/artifacts/images/css_fractal_candidate_logical_expanded_rust.png")]
    output: PathBuf,
}

// GF(2) rank via Gaussian elimination
fn gf2_rank(matrix: &mut [Vec<u8>]) -> usize {
    if matrix.is_empty() {
        return 0;
    }
    let m = matrix.len();
    let n = matrix[0].len();
    let mut r = 0;

    for c in 0..n {
        let mut piv = None;
        #[allow(clippy::needless_range_loop)]
        for i in r..m {
            if matrix[i][c] == 1 {
                piv = Some(i);
                break;
            }
        }
        
        if let Some(pivot_row) = piv {
            if pivot_row != r {
                matrix.swap(r, pivot_row);
            }
            #[allow(clippy::needless_range_loop)]
            for i in 0..m {
                if i != r && matrix[i][c] == 1 {
                    for col in 0..n {
                        matrix[i][col] ^= matrix[r][col];
                    }
                }
            }
            r += 1;
            if r == m {
                break;
            }
        }
    }
    r
}

fn main() -> Result<()> {
    let args = Args::parse();
    println!("Computing GF(2) Ranks for Fractal Code...");

    let l = 33;
    let n_qubits = l * l;
    let mut s = vec![vec![0u8; l]; l];

    // Seed Rule-90
    s[0][l / 2] = 1;
    for r in 1..l {
        for c in 0..l {
            let left_idx = if c == 0 { l - 1 } else { c - 1 };
            let right_idx = if c == l - 1 { 0 } else { c + 1 };
            let left = s[r - 1][left_idx];
            let right = s[r - 1][right_idx];
            s[r][c] = left ^ right;
        }
    }

    let mut x_parity = vec![vec![0u8; l]; l];
    let mut z_parity = vec![vec![0u8; l]; l];
    for r in 0..l {
        for c in 0..l {
            if (r + c) % 2 == 0 {
                x_parity[r][c] = 1;
            } else {
                z_parity[r][c] = 1;
            }
        }
    }

    let translate = |mask: &[Vec<u8>], tx: usize, ty: usize| -> Vec<Vec<u8>> {
        let mut out = vec![vec![0u8; l]; l];
        #[allow(clippy::needless_range_loop)]
        for r in 0..l {
            for c in 0..l {
                let src_r = (r + l - ty % l) % l;
                let src_c = (c + l - tx % l) % l;
                out[r][c] = mask[src_r][src_c];
            }
        }
        out
    };

    let x_shifts = vec![(0,0),(2,0),(0,2),(2,2),(4,2),(2,4),(6,0),(0,6),(6,4),(4,6),(8,2),(2,8)];
    let z_shifts = vec![(2,0),(0,2),(4,0),(0,4),(6,2),(2,6),(4,4),(8,0),(0,8),(6,6),(8,4),(4,8)];

    let flat = |mask: &[Vec<u8>]| -> Vec<u8> {
        let mut out = Vec::with_capacity(n_qubits);
        #[allow(clippy::needless_range_loop)]
        for r in 0..l {
            for c in 0..l {
                out.push(mask[r][c] % 2);
            }
        }
        out
    };

    let mut hx_orig = Vec::new();
    for &(tx, ty) in &x_shifts {
        let translated = translate(&s, tx, ty);
        let mut mask = vec![vec![0u8; l]; l];
        for r in 0..l {
            for c in 0..l {
                mask[r][c] = translated[r][c] & x_parity[r][c];
            }
        }
        hx_orig.push(flat(&mask));
    }

    let mut hz = Vec::new();
    for &(tx, ty) in &z_shifts {
        let translated = translate(&s, tx, ty);
        let mut mask = vec![vec![0u8; l]; l];
        for r in 0..l {
            for c in 0..l {
                mask[r][c] = translated[r][c] & z_parity[r][c];
            }
        }
        hz.push(flat(&mask));
    }

    let mut hx_working = hx_orig.clone();
    let mut hz_working = hz.clone();
    let rank_hx = gf2_rank(&mut hx_working);
    let rank_hz = gf2_rank(&mut hz_working);
    let k = n_qubits - rank_hx - rank_hz;

    println!("L={}, n={}, rank(HX)={}, rank(HZ)={}, k={}", l, n_qubits, rank_hx, rank_hz, k);

    let mut best: Option<(usize, usize, usize, Vec<Vec<u8>>)> = None;

    for tx in (0..l).step_by(2) {
        for ty in (0..l).step_by(2) {
            let translated = translate(&s, tx, ty);
            let mut xmask = vec![vec![0u8; l]; l];
            for r in 0..l {
                for c in 0..l {
                    xmask[r][c] = translated[r][c] & x_parity[r][c];
                }
            }
            let xrow = flat(&xmask);

            // must commute with all Z stabilizers
            let mut commutes = true;
            for zrow in &hz {
                let mut dot = 0;
                for i in 0..n_qubits {
                    dot ^= xrow[i] & zrow[i];
                }
                if dot != 0 {
                    commutes = false;
                    break;
                }
            }
            if !commutes { continue; }

            // must be independent of H_X
            let mut hx_ext = hx_orig.clone();
            hx_ext.push(xrow.clone());
            if gf2_rank(&mut hx_ext) == rank_hx {
                continue;
            }

            let weight: usize = xrow.iter().map(|&x| x as usize).sum();
            if best.is_none() || weight < best.as_ref().unwrap().0 {
                best = Some((weight, tx, ty, xmask));
            }
        }
    }

    if let Some((w_best, tx_best, ty_best, xlog)) = best {
        println!("Found X-Logical: w={}, shift=({},{})", w_best, tx_best, ty_best);

        let cell_size = 20;
        let img_size = (l * cell_size) as u32;
        let mut img = ImageBuffer::new(img_size, img_size);

        let bg_color = Rgb([13u8, 15u8, 20u8]);
        let blue = Rgb([97u8, 166u8, 245u8]);

        #[allow(clippy::needless_range_loop)]
        for r in 0..l {
            for c in 0..l {
                let color = if xlog[r][c] == 1 { blue } else { bg_color };
                for dy in 0..cell_size {
                    for dx in 0..cell_size {
                        let px = (c * cell_size + dx) as u32;
                        let py = (r * cell_size + dy) as u32;
                        img.put_pixel(px, py, color);
                    }
                }
            }
        }

        if let Some(parent) = args.output.parent() {
            std::fs::create_dir_all(parent)?;
        }
        img.save(&args.output)?;
        println!("Saved: {}", args.output.display());
    } else {
        println!("No candidate X-logical found in this restricted family.");
    }

    Ok(())
}
